pkg/apis/azure/validation/infrastructure.go

```diff
@@ -62,28 +115,159 @@ func ValidateInfrastructureConfig(infra *apisazure.InfrastructureConfig, nodesCI
 	}
 
-	allErrs = append(allErrs, validateVnetConfig(infra.Networks.VNet, infra.ResourceGroup, workerCIDR, nodes, pods, services, networksPath.Child("vnet"))...)
-	allErrs = append(allErrs, validateNatGatewayConfig(infra.Networks.NatGateway, infra.Zoned, hasVmoAlphaAnnotation, networksPath.Child("natGateway"))...)
+	return allErrs
+}
 
-	if infra.Identity != nil && (infra.Identity.Name == "" || infra.Identity.ResourceGroup == "") {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("identity"), infra.Identity, "specifying an identity requires the name of the identity and the resource group which hosts the identity"))
+func validateNetworkConfig(
+	infra *apisazure.InfrastructureConfig,
+	nodes cidrvalidation.CIDR,
+	pods cidrvalidation.CIDR,
+	services cidrvalidation.CIDR,
+	hasVmoAlphaAnnotation bool,
+	fldPath *field.Path,
+) field.ErrorList {
+
+	var (
+		config       = infra.Networks
+		networksPath = fldPath.Child("networks")
+		allErrs      = field.ErrorList{}
+		workersPath  = networksPath.Child("workers")
+		zonesPath    = networksPath.Child("zones")
+		vNetPath     = networksPath.Child("vnet")
+		workerCIDR   cidrvalidation.CIDR
+	)
+
+	// forbid not setting at least one of {workers, zones}
+	if config.Workers == nil && len(config.Zones) == 0 {
+		allErrs = append(allErrs, field.Forbidden(workersPath, "either workers or zones must be specified"))
+		return allErrs
+	}
+
+	// forbid setting both of {workers, zones}
+	if config.Workers != nil && len(config.Zones) > 0 {
+		allErrs = append(allErrs, field.Forbidden(workersPath, "workers and zones cannot be specified in parallel"))
+		return allErrs
+	}
+
+	if helper.IsUsingSingleSubnetLayout(infra) {
+		workerCIDR = cidrvalidation.NewCIDR(*config.Workers, workersPath)
+		allErrs = append(allErrs, cidrvalidation.ValidateCIDRParse(workerCIDR)...)
+		allErrs = append(allErrs, cidrvalidation.ValidateCIDRIsCanonical(workersPath, *config.Workers)...)
+
+		if nodes != nil {
+			allErrs = append(allErrs, nodes.ValidateSubset(workerCIDR)...)
+		}
+
+		allErrs = append(allErrs, validateNatGatewayConfig(config.NatGateway, infra.Zoned, hasVmoAlphaAnnotation, networksPath.Child("natGateway"))...)
+	} else {
+		if !infra.Zoned {
+			allErrs = append(allErrs, field.Forbidden(zonesPath, "cannot specify zones in an non-zonal cluster"))
+		}
+
+		if config.NatGateway != nil {
+			allErrs = append(allErrs, field.Forbidden(workersPath, "natGateway cannot be specified when workers field is missing"))
+		}
+		if len(config.ServiceEndpoints) > 0 {
+			allErrs = append(allErrs, field.Forbidden(workersPath, "serviceEndpoints cannot be specified when workers field is missing"))
+		}
+
+		allErrs = append(allErrs, validateZones(config.Zones, nodes, pods, services, zonesPath)...)
 	}
 
+	allErrs = append(allErrs, validateVnetConfig(&config, infra.ResourceGroup, workerCIDR, nodes, pods, services, zonesPath, vNetPath)...)
```

Comment:
Can you move this before the `if else` block?
Then we could eliminate the `if else` by returning after the `helper.IsUsingSingleSubnetLayout` if block is completed.